### PR TITLE
[Sprint 64] Feature/2105 - Schedule F Aggregation

### PR DESCRIPTION
**Pull request by Elaine-Krauss-TCG**
Ticket link: [here](https://fecgov.atlassian.net/browse/FECFILE-2105)
  
Related PRs: [API](https://github.com/fecgov/fecfile-web-api/pull/1480), [App](https://github.com/fecgov/fecfile-web-app/pull/2719)

### CODE REVIEW

File: django-backend/fecfiler/transactions/views.py [hunk 1]
```diff
@@ -396,6 +438,26 @@ def save_transaction(self, transaction_data, request):
         delete_carried_forward_loans_if_needed(transaction_instance, committee_id)
         delete_carried_forward_debts_if_needed(transaction_instance, committee_id)
 
+        # Handle changing general_election_year for Schedule F transactions
+        if (
```
Comment (toddlees) [inline django-backend/fecfiler/transactions/views.py:442]: save_transaction is getting pretty big.  I know it would mean passing in more info to ```recalculate_schedule_f_aggregates```, but if we did, could we not figure out ALL of the transactions that need to be recalculated in one place rather than spread across 3 spots in save_transaction?

File: django-backend/fecfiler/transactions/views.py [hunk 2]
```diff
@@ -508,6 +582,50 @@ def remove_transaction_from_report(self, request):
...
+            transaction.schedule_f.general_election_year
+        )
+
+        to_update = shared_entity_transactions.filter(
+            Q(id=transaction.id)
+            | Q(date__gt=transaction.date)
+            | Q(Q(date=transaction.date) & Q(created__gt=transaction.created))
+        ).order_by("date", "created")
+
+        previous_transaction = previous_transactions.first()
```
Comment (toddlees) [inline django-backend/fecfiler/transactions/views.py:613]: i believe this snippet is duplicate work

**Review by toddlees — CHANGES_REQUESTED** [review]

File: django-backend/fecfiler/transactions/utils.py [hunk 3]
```diff
@@ -37,3 +38,48 @@ def add_location(data, contact: Contact, identifier: str):
...
+            Q(contact_2__candidate_office=office),
+            Q(contact_2__candidate_state=state),
+            Q(contact_2__candidate_district=district),
+        )
+
+    original_transaction = None
+    if transaction_id:
+        original_transaction = queryset.get(id=transaction_id)
+        query = query.filter(
+            Q(created__lt=original_transaction.created) | ~Q(date=date)
```
Comment (dheitzer) [inline django-backend/fecfiler/transactions/utils.py:81]: should this be ~Q(date >= date) to make sure we're always getting previous transactions? Also, couldn't transactions with older created dates still have newer dates?

File: django-backend/fecfiler/transactions/views.py [hunk 4]
```diff
@@ -509,6 +672,59 @@ def remove_transaction_from_report(self, request):
         transaction.reports.remove(report)
         return Response("Transaction removed from report")
 
+    def process_general_election_year_aggregates(self, transaction_instance):
```
Comment (dheitzer) [inline django-backend/fecfiler/transactions/views.py:675]: I'm wondering if we can name this function more descriptive - maybe update_previous_transactions_aggregates()?

File: django-backend/fecfiler/transactions/views.py [hunk 5]
```diff
@@ -509,6 +672,59 @@ def remove_transaction_from_report(self, request):
...
+                    previous_transaction.schedule_f.aggregate_general_elec_expended
+                )
+
+            trans.schedule_f.aggregate_general_elec_expended = (
+                trans.schedule_f.expenditure_amount + previous_aggregate
+            )
+            updated_schedule_fs.append(trans.schedule_f)
+            previous_transaction = trans
+
+        ScheduleF.objects.bulk_update(
```
Comment (dheitzer) [inline django-backend/fecfiler/transactions/views.py:723]: I'm wondering if we might want to set the batch_size parameter in case we end up with an excessive number of previous transactions.  Thoughts?
https://docs.djangoproject.com/en/5.2/ref/models/querysets/#bulk-update

File: django-backend/fecfiler/transactions/views.py [hunk 6]
```diff
@@ -509,6 +672,59 @@ def remove_transaction_from_report(self, request):
...
+            transaction.id,
+            transaction.date,
+            transaction.aggregation_group,
+            None,
+            transaction.contact_2.id,
+            None,
+            transaction.schedule_f.general_election_year
+        )
+
+        to_update = shared_entity_transactions.filter(
```
Comment (dheitzer) [inline django-backend/fecfiler/transactions/views.py:702]: Should this be including submitted transactions and such?

File: django-backend/fecfiler/transactions/views.py (hunk 5 above)
Comment (Elaine-Krauss-TCG) [reply django-backend/fecfiler/transactions/views.py:723]: It might be unnecessary because we're just altering one column, but I'm open to the idea.  Any thoughts on what a good size for the batch size would be?
image

File: django-backend/fecfiler/transactions/utils.py [hunk 7]
```diff
@@ -37,3 +38,48 @@ def add_location(data, contact: Contact, identifier: str):
     data[f"{identifier}_city"] = contact.city
     data[f"{identifier}_state"] = contact.state
     data[f"{identifier}_zip"] = contact.zip
+
+
+def filter_for_previous_transactions(
```
[the remark this answers is not in the document]
Comment (Elaine-Krauss-TCG) [reply django-backend/fecfiler/transactions/utils.py:43]: Sounds good; I'm working on a commit for that

File: django-backend/fecfiler/transactions/utils.py (hunk 3 above)
Comment (Elaine-Krauss-TCG) [reply django-backend/fecfiler/transactions/utils.py:81]: I'm pretty convinced that I've got all the logic working correctly, and I think I have a pretty comprehensive set of unit and E2E tests, but if you see even a chance that there's room for an edge-case, then we should definitely take that seriously.  I'll take another look at my unit tests and make sure that there are tests for the aggregation logic around creation order

Comment (toddlees) [reply django-backend/fecfiler/transactions/utils.py:81]: I think it's correct @Elaine-Krauss-TCG maybe we can just add a comment explaining the purpose of this one filter.  I think the confusion is in thinking it's meant to be the whole filter, when really it's just addressing a subset of the transactions that are filtered in the whole function (ones that are the same date but created before the original)

**Review by toddlees — APPROVED** [review]
Anything you can address in the comments is great. Overall, this is in a good place

File: django-backend/fecfiler/transactions/utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Elaine-Krauss-TCG) [reply django-backend/fecfiler/transactions/utils.py]: I broke it down a lot, and hopefully between comments and the names of helper functions, it should be a lot easier to read through.  Additionally, by having helper functions return Q objects, I was able to get the newly renamed `filter_queryset_for_previous_transactions_in_aggregation()` to only make one `.filter()` call

**Review by danguyf — APPROVED** [review]
(chef's kiss emoji)

**Review by dheitzer — APPROVED** [review]

File: django-backend/fecfiler/transactions/utils.py (hunk 3 above)
Comment (dheitzer) [reply django-backend/fecfiler/transactions/utils.py]: yes @toddlees I think I misunderstood this filter - thanks for all the checking and clarification @toddlees @Elaine-Krauss-TCG
